[TensorExpr] Implement shape inference for TE.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42387 [TensorExpr] Support shape inference in TE for aten::cat.
* #42202 [TensorExpr] Make tensorOrConstant non-templatized function.
* **#41451 [TensorExpr] Implement shape inference for TE.**
* #42201 [TensorExpr] Explicitly cast to bool results of comparison ops in kernel.cpp.
* #42196 [TensorExpr] Add debug dumps to kernel.cpp.

Since TE operates on a limited subset of ops with a well-defined
semantics, we can easily infer shapes of intermediate and output tensors
given shapes of the inputs.

There is a couple of ops that are not yet supported in the shape
inference, once we add them we could relax the shape info requirements
in the TE fuser: currently it requires all values in the fusion group to
have shapes known and we can change it to only inputs.

Differential Revision: [D22543470](https://our.internmc.facebook.com/intern/diff/D22543470)

Commit:
[TensorExpr] Implement shape inference for TE.

Since TE operates on a limited subset of ops with a well-defined
semantics, we can easily infer shapes of intermediate and output tensors
given shapes of the inputs.

There is a couple of ops that are not yet supported in the shape
inference, once we add them we could relax the shape info requirements
in the TE fuser: currently it requires all values in the fusion group to
have shapes known and we can change it to only inputs.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement shape inference for TE."


Since TE operates on a limited subset of ops with a well-defined
semantics, we can easily infer shapes of intermediate and output tensors
given shapes of the inputs.

There is a couple of ops that are not yet supported in the shape
inference, once we add them we could relax the shape info requirements
in the TE fuser: currently it requires all values in the fusion group to
have shapes known and we can change it to only inputs.

Differential Revision: [D22543470](https://our.internmc.facebook.com/intern/diff/D22543470)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement shape inference for TE."


Since TE operates on a limited subset of ops with a well-defined
semantics, we can easily infer shapes of intermediate and output tensors
given shapes of the inputs.

There is a couple of ops that are not yet supported in the shape
inference, once we add them we could relax the shape info requirements
in the TE fuser: currently it requires all values in the fusion group to
have shapes known and we can change it to only inputs.

Differential Revision: [D22543470](https://our.internmc.facebook.com/intern/diff/D22543470)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement shape inference for TE."


Since TE operates on a limited subset of ops with a well-defined
semantics, we can easily infer shapes of intermediate and output tensors
given shapes of the inputs.

There is a couple of ops that are not yet supported in the shape
inference, once we add them we could relax the shape info requirements
in the TE fuser: currently it requires all values in the fusion group to
have shapes known and we can change it to only inputs.

Differential Revision: [D22543470](https://our.internmc.facebook.com/intern/diff/D22543470)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement shape inference for TE."


Since TE operates on a limited subset of ops with a well-defined
semantics, we can easily infer shapes of intermediate and output tensors
given shapes of the inputs.

There is a couple of ops that are not yet supported in the shape
inference, once we add them we could relax the shape info requirements
in the TE fuser: currently it requires all values in the fusion group to
have shapes known and we can change it to only inputs.

Differential Revision: [D22543470](https://our.internmc.facebook.com/intern/diff/D22543470)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement shape inference for TE."


Since TE operates on a limited subset of ops with a well-defined
semantics, we can easily infer shapes of intermediate and output tensors
given shapes of the inputs.

There is a couple of ops that are not yet supported in the shape
inference, once we add them we could relax the shape info requirements
in the TE fuser: currently it requires all values in the fusion group to
have shapes known and we can change it to only inputs.

Differential Revision: [D22543470](https://our.internmc.facebook.com/intern/diff/D22543470)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Implement shape inference for TE."


Since TE operates on a limited subset of ops with a well-defined
semantics, we can easily infer shapes of intermediate and output tensors
given shapes of the inputs.

There is a couple of ops that are not yet supported in the shape
inference, once we add them we could relax the shape info requirements
in the TE fuser: currently it requires all values in the fusion group to
have shapes known and we can change it to only inputs.

Differential Revision: [D22543470](https://our.internmc.facebook.com/intern/diff/D22543470)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -189,21 +328,33 @@ static bool isOne(ExprHandle e) {
-static std::pair<std::vector<ExprHandle>, bool> broadcastShapes(
+std::vector<ExprHandle> TensorExprKernel::broadcastShapes(
+    std::vector<std::vector<ExprHandle>> shapes) {
+  bool seenBroadcast = false;
```
unused

eellison reviewed (approved):
Accepting to unblock. We're going to need to think up a more comprehensive testing strategy before enabling TE Fuser on in master, which should hopefully flesh out any issues here. I also suspect a lot of this will need to be rethought and revisited if/when we enable TE with unknown dimensions.

@Krovatkin maybe you can take a look as well

Commit:
Update on "[TensorExpr] Implement shape inference for TE."


Since TE operates on a limited subset of ops with a well-defined
semantics, we can easily infer shapes of intermediate and output tensors
given shapes of the inputs.

There is a couple of ops that are not yet supported in the shape
inference, once we add them we could relax the shape info requirements
in the TE fuser: currently it requires all values in the fusion group to
have shapes known and we can change it to only inputs.

Differential Revision: [D22543470](https://our.internmc.facebook.com/intern/diff/D22543470)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Implement shape inference for TE."


Since TE operates on a limited subset of ops with a well-defined
semantics, we can easily infer shapes of intermediate and output tensors
given shapes of the inputs.

There is a couple of ops that are not yet supported in the shape
inference, once we add them we could relax the shape info requirements
in the TE fuser: currently it requires all values in the fusion group to
have shapes known and we can change it to only inputs.

Differential Revision: [D22543470](https://our.internmc.facebook.com/intern/diff/D22543470)

[ghstack-poisoned]
